gdbm/doc/gdbm_dump.1
.\" This file is part of GDBM.
.\" Copyright (C) 2013, 2017-2020 Free Software Foundation, Inc.
.\"
.\" GDBM is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 3, or (at your option)
.\" any later version.
.\"
.\" GDBM is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with GDBM. If not, see <http://www.gnu.org/licenses/>. */
.TH GDBM_DUMP 1 "May 8, 2013" "GDBM" "GDBM User Reference"
.SH NAME
gdbm_dump \- dump a GDBM database to a file
.SH SYNOPSIS
\fBgdbm_dump\fR [\fB\-H \fIFMT\fR] [\fB\-\-format\fR=\fIFMT\fR] \fIDB_FILE\fR [\fIFILE\fR]
.sp
\fBgdbm_dump\fR [\fB\-Vh\fR] [\fB\-\-help\fR] [\fB\-\-usage\fR] [\fB\-\-version\fR]
.SH DESCRIPTION
The
.B gdbm_dump
utility creates a dump of the specified
.BR gdbm (3)
database file.  The name for the output dump file is supplied by the
second argument (\fIFILE\fR).  If not specified, the output goes to
the standard error.
.PP
The created dump can be given as argument to the
.BR gdbm_load (1)
utility in order to re-create an exact copy of the \fIDB_FILE\fR.
.SH OPTIONS
.TP
\fB\-H\fR, \fB\-\-format\fR=\fIFMT\fR
Select dump format.  The value \fBbinary\fR (or \fB0\fR) instructs
.B gdbm_dump
to produce a binary dump, compatible with earlier
.B gdbm
versions (up to version 1.9).  The value \fBascii\fR (or \fB1\fR)
instructs it to create an ASCII dump (this is the default).  The
latter is preferred because, apart from the actual data, it also
contains meta-information which will allow
.BR gdbm_load (1)
to recreate an exact copy of the file.
.TP
\fB\-h\fR, \fB\-\-help\fR
Print a short usage summary.
.TP
\fB\-\-usage\fR
Print a list of available options.
.TP
\fB\-V\fR, \fB\-\-version\fR
Print program version
.SH "SEE ALSO"
.BR gdbm_load (1),
.BR gdbmtool (1),
.BR gdbm (3).
.PP
For a detailed description of
.B gdbm_dump
and other
.B gdbm
utilities, refer to the \fBGDBM Manual\fR available in
Texinfo format.  To access it, run:

  \fBinfo gdbm\fR

.SH "REPORTING BUGS"
Report bugs to <bug\-gdbm@gnu.org>.
.SH COPYRIGHT
Copyright \(co 2013 Free Software Foundation, Inc
.br
.na
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
.br
.ad
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
.\" Local variables:
.\" eval: (add-hook 'write-file-hooks 'time-stamp)
.\" time-stamp-start: ".TH GDBM[A-Z_-]* 1 \""
.\" time-stamp-format: "%:B %:d, %:y"
.\" time-stamp-end: "\""
.\" time-stamp-line-limit: 20
.\" end:


gdbm/NEWS
GNU dbm NEWS -- history of user-visible changes. 2020-12-23
Copyright (C) 1990-2020 Free Software Foundation, Inc.
See the end of file for copying conditions.

Please send gdbm bug reports to <bug-gdbm@gnu.org>.

Version 1.19 - 2020-12-23

* Pre-read the memory mapped regions on systems that support it.

This speeds up operations on big databases.

* gdbmtool: tagged initialization of structured data

Initializers for structured data can be given in tagged form, e.g.:

   store somekey { status=2, id={a,u,x}, name="foo" }

* Bugfixes:

** Preserve locking type during database reorganization


Version 1.18.1 - 2018-10-27

* Fix debian bug 910911

This restores backward compatibility with databases created by version
1.8 (and some later versions, with mmapped I/O disabled).
See https://bugs.debian.org/910911 for a detailed description.


Version 1.18 - 2018-08-21

* Bugfixes:

** Fix directory entry validation
See https://puszcza.gnu.org.ua/bugs/?402

Directory validation function would falsely report corruption after
several directory reallocations.

** Fix improper use of GDBM_DEBUG macro
See https://puszcza.gnu.org.ua/bugs/?401

** Fix spurious error from gdbm_dump and gdbm_export

The functions incorrectly reported as error the GDBM_ITEM_NOT_FOUND
status, which is reported when upon normal termination of iteration
over the database keys.

** Make sure gdbm_sync always returns a meaningful value
See https://puszcza.gnu.org.ua/bugs/?400

Version 1.17 - 2018-07-30

* int gdbm_close and gdbm_sync

Both functions now return 0 on success. On error, they return -1
and set gdbm_errno and errno to the appropriate error codes.

* Fix bug in gdbm_dump function

The function did not clear the GDBM_ITEM_NOT_FOUND error marker, which
is set upon its normal termination.


Version 1.16 - 2018-06-27

* Maintain sorting order of the available block list after coalescing

This is a very long-standing bug, dated back to the time when optional
coalescing of available blocks with the newly released block was
introduced. Merging the released block with an adjacent block of
available space would clobber the sorting order of the available list.
The bug was discovered due to strict database consistency checks,
introduced in version 1.15.

The fix introduced in this version will silently restore the proper
sorting order of available block list before its use.

* Improve block merging algorithm

New implementation of block merging algorithm will correctly handle
both left and right-adjacent blocks.


Version 1.15 - 2018-06-15

* Extensive database consistency checking

GDBM tries to detect inconsistencies in input database files as
early as possible. When an inconcistency is detected, a helpful
diagnostics is returned and the database is marked as needing
recovery. From this moment on, any GDBM function trying to access
the database will immediately return error code (instead of
eventually segfaulting as previous versions did). In order to
reconstruct the database and return it to healthy state, the
gdbm_recover function should be used.

Special thanks to Lionel Debroux and Craig Young for investing
their time and efforts in testing and providing valuable feedback.

* Improved error checking

* Removed gdbm-1.8.3 compatibility layer

* Commands can be given to gdbmtool in the command line

The syntax is:

  gdbmtool DBNAME COMMAND [ARGS...]

Multiple commands are separated by semicolon (take care to escape it),
e.g.:

  gdbmtool t.db count\; avail

* Fixed data conversion bugs in storing structured keys or content

* New member in the gdbm_recovery structure: duplicate_keys.

Upon return from gdbm_recover, this member holds the number of keys
that has not been recovered, because the same key had already been stored
in the database. The actual number of stored keys is thus
recovered_keys - duplicate_keys.

* New error codes.

  GDBM_BAD_BUCKET      "Malformed bucket header"
  GDBM_BAD_HEADER      "Malformed database file header"
  GDBM_BAD_AVAIL       "Malformed avail_block"
  GDBM_BAD_HASH_TABLE  "Malformed hash table"
  GDBM_BAD_DIR_ENTRY   "Invalid directory entry"


Version 1.14.1 - 2018-01-03

* Increment soname current version number.


Version 1.14 - 2018-01-01

* Make sure created databases are byte-for-byte reproducible

This fixes two longstanding bugs: (1) when allocating database file
header blocks, the unused memory is filled with zeroes; (2) when expanding
a mmapped memory area, the added extent is filled with zeroes.

* Fix build with --enable-gdbm-export

* Make gdbm_error global variable thread safe.

* Fix possible segmentation violation in gdbm_setopt

* Fix handling of group headers in --help output.


Version 1.13 - 2017-03-11

* gdbm_fetch, gdbm_firstkey, and gdbm_nextkey behavior

If the requested key was not found, these functions return datum with
dptr pointing to NULL and set gdbm_errno to GDBM_ITEM_NOT_FOUND (in
prior releases, gdbm_errno was set to GDBM_NO_ERROR),

If an error occurred, dptr is set to NULL, and gdbm_errno to
the error code.

In any case gdbm_errno is guaranteed to have meaningful value upon
return.

* Error handling

In previous versions of GDBM, fatal errors (such as write error while
storing the key/data pair or bucket) caused immediate termination of
the program via call to exit(3).  This is no longer the case.

Starting from this version, if a fatal error occurrs while modifying
the database file, that database is marked as needing recovery and
gdbm_errno is set to GDBM_NEED_RECOVERY.  Calls to any GDBM functions,
except gdbm_recover, will then return immediately with the same error
code.

The function gdbm_recover examines the database file and fixes
eventual inconsistencies.  Upon successful return it clears the error
state and makes the database operational again.

For backward compatibility, the fatal_func argument to gdbm_open is
retained and its functionality is not changed.  If it is not NULL, the

gdbm/src/gdbm_dump.c
/* This file is part of GDBM, the GNU data base manager.
   Copyright (C) 2011, 2013, 2017-2020 Free Software Foundation, Inc.

   GDBM is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 3, or (at your option)
   any later version.

   GDBM is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with GDBM. If not, see <http://www.gnu.org/licenses/>.   */

# include "autoconf.h"
# include "gdbm.h"
# include "gdbmapp.h"
# include "gdbmdefs.h"

char *parseopt_program_doc = N_("dump a GDBM database to a file");
char *parseopt_program_args = N_("DB_FILE [FILE]");
struct gdbm_option optab[] = {
  { 'H', "format", "binary|ascii|0|1", N_("select dump format") },
  { 0 }
};

int format = GDBM_DUMP_FMT_ASCII;

int
main (int argc, char **argv)
{
  GDBM_FILE dbf;
  int rc, opt;
  char *dbname, *filename;
  FILE *fp;

#ifdef HAVE_SETLOCALE
  setlocale (LC_ALL, "");
#endif
  bindtextdomain (PACKAGE, LOCALEDIR);
  textdomain (PACKAGE);

  set_progname (argv[0]);

  for (opt = parseopt_first (argc, argv, optab);
       opt != EOF;
       opt = parseopt_next ())
    {
    switch (opt)
      {
      case 'H':
	if (strcmp (optarg, "binary") == 0)
	  format = GDBM_DUMP_FMT_BINARY;
	else if (strcmp (optarg, "ascii") == 0)
	  format = GDBM_DUMP_FMT_ASCII;
	else
	  {
	    format = atoi (optarg);
	    switch (format)
	      {
	      case GDBM_DUMP_FMT_BINARY:
	      case GDBM_DUMP_FMT_ASCII:
		break;
	      default:
		error (_("unknown dump format"));
		exit (EXIT_USAGE);
	      }
	  }
	break;
	
      default:
	error (_("unknown option"));
	exit (EXIT_USAGE);
      }
    }

  argc -= optind;
  argv += optind;

  if (argc == 0)
    {
      parseopt_print_help ();
      exit (EXIT_OK);
    }

  if (argc > 2)
    {
      error (_("too many arguments; try `%s -h' for more info"), progname);
      exit (EXIT_USAGE);
    }
  
  dbname = argv[0];
  if (argc == 2)
    filename = argv[1];
  else
    filename = NULL;

  if (!filename || strcmp (filename, "-") == 0)
    {
      filename = "<stdout>";
      fp = stdout;
    }
  else
    {
      fp = fopen (filename, "w");
      if (!fp)
	{
	  sys_perror (errno, _("cannot open %s"), filename);
	  exit (EXIT_FATAL);
	}
    }

  dbf = gdbm_open (dbname, 0, GDBM_READER, 0600, NULL);
  if (!dbf)
    {
      gdbm_perror (_("gdbm_open failed"));
      exit (EXIT_FATAL);
    }

  rc = gdbm_dump_to_file (dbf, fp, format);
  if (rc)
    {
      gdbm_perror (_("dump error"), filename);
    }
  
  gdbm_close (dbf);

  exit (rc == GDBM_NO_ERROR ? EXIT_OK : EXIT_FATAL);
}
  
